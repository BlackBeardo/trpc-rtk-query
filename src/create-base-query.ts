import { type BaseQueryApi, type BaseQueryFn } from "@reduxjs/toolkit/query/react";
import {
  type CreateTRPCClientOptions,
  type CreateTRPCProxyClient,
  TRPCClientError,
  type TRPCRequestOptions,
  TRPCUntypedClient,
  createTRPCUntypedClient,
  getUntypedClient,
} from "@trpc/client";
import { type AnyRouter, TRPCError } from "@trpc/server";
import { getHTTPStatusCodeFromError } from "@trpc/server/http";

/* Errors baseQuery can return. Follows the conventions of RTK query's fetchBaseQuery */
export type TRPCBaseQueryError =
  | {
      /**
       * * `"TRPC_CLIENT_ERROR"`:
       *   An error that happened on trpc client. Original error is stringified in error
       *   attribute.
       **/
      data?: undefined;
      error: string;
      message: string;
      name: string;
      status: "TRPC_CLIENT_ERROR";
    }
  | {
      /**
       * * `"TRPC_ERROR"`:
       *   An error that was returned by trpc backend. Original error is stringified in
       *   error attribute.
       **/
      data?: undefined;
      error: string;
      message: string;
      name: string;
      status: "TRPC_ERROR";
      statusCode: number;
    }
  | {
      /**
       * * `"UNKNOWN_ERROR"`:
       *   A unknown error type that captures error not wrapped by trpc. Original error
       *   is stringified in error attribute
       **/
      data?: unknown;
      error: string;
      status: "UNKNOWN_ERROR";
    };

/**
 * Client specific options when creating trpc api. You can either
 *  - pass in already created proxy client
 *  - pass in client options, so client is created for you
 *  - pass in getClient -function, that has gets access to BaseQueryApi from RTK. This
 *    is useful e.g. when you need to access redux store when creating the client.
 **/
export type CreateTRPCApiClientOptions<TRouter extends AnyRouter> =
  | {
      client: CreateTRPCProxyClient<TRouter>;
      clientOptions?: never;
      getClient?: never;
    }
  | {
      client?: never;
      clientOptions: CreateTRPCClientOptions<TRouter>;
      getClient?: never;
    }
  | {
      client?: never;
      clientOptions?: never;
      getClient: (
        baseQueryApi: BaseQueryApi,
      ) => Promise<CreateTRPCProxyClient<TRouter>>;
    };

/** Helper type to differentiate between whether client is ready to be used or should
 *  be grabbed from callback
 *  @internal
 **/
type ClientStatus<TRouter extends AnyRouter> =
  | {
      client: TRPCUntypedClient<TRouter>;
      ready: true;
    }
  | {
      getClient: NonNullable<CreateTRPCApiClientOptions<TRouter>["getClient"]>;
      ready: false;
    };

/** Resolve how to get the trpc client
 * @internal
 **/
const resolveClientStatus = <TRouter extends AnyRouter>(
  createTRPCApiClientOptions: CreateTRPCApiClientOptions<TRouter>,
): ClientStatus<TRouter> => {
  if ("client" in createTRPCApiClientOptions) {
    return {
      client: getUntypedClient<TRouter>(createTRPCApiClientOptions.client),
      ready: true,
    };
  } else if ("clientOptions" in createTRPCApiClientOptions) {
    return {
      client: createTRPCUntypedClient<TRouter>(
        createTRPCApiClientOptions.clientOptions,
      ),
      ready: true,
    };
  }
  return {
    getClient: createTRPCApiClientOptions.getClient,
    ready: false,
  };
};

/** Typings for base query that (not previously existing) api created with
 * createTRPCApi uses.
 **/
export type BaseQueryForTRPCClient = BaseQueryFn<
  // Arguments for baseQuery. Should be used when injecting endpoints
  {
    procedureArguments: unknown;
    procedurePath: string;
    procedureType: "mutation" | "query";
  },
  // Result type. Endpoint definitions are typed manually instead of deriving them from
  // this type, so it's enough that this is unknown.
  unknown,
  // Use typed errors
  TRPCBaseQueryError,
  // Allow request options (e.g context and signal) to be passed in via extra options
  TRPCRequestOptions,
  // No meta is returned
  never
>;

/** Create a base query that uses trpc client under the hood.
 **/
export const createBaseQueryForTRPCClient = <TRouter extends AnyRouter>(
  createTRPCApiClientOptions: CreateTRPCApiClientOptions<TRouter>,
): BaseQueryForTRPCClient => {
  const clientStatus = resolveClientStatus(createTRPCApiClientOptions);

  return async (baseQueryArguments, baseQueryApi, extraOptions) => {
    try {
      const { procedureArguments, procedurePath, procedureType } = baseQueryArguments;
      const resolvedClient = clientStatus.ready
        ? clientStatus.client
        : getUntypedClient<TRouter>(await clientStatus.getClient(baseQueryApi));
      return {
        data: await resolvedClient[procedureType](
          procedurePath,
          procedureArguments,
          extraOptions,
        ),
      };
    } catch (error) {
      let properlyShapedError: {
        error: TRPCBaseQueryError;
      };
      if (error instanceof TRPCClientError) {
        properlyShapedError = {
          error: {
            error: String(error),
            message: error.message,
            name: error.name,
            status: "TRPC_CLIENT_ERROR",
          },
        };
      } else if (error instanceof TRPCError) {
        properlyShapedError = {
          error: {
            error: String(error),
            message: error.message,
            name: error.name,
            status: "TRPC_ERROR",
            statusCode: getHTTPStatusCodeFromError(error),
          },
        };
      } else {
        properlyShapedError = {
          error: {
            error: String(error),
            status: "UNKNOWN_ERROR",
          },
        };
      }
      return properlyShapedError;
    }
  };
};
