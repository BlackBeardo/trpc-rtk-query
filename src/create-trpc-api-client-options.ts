import { type BaseQueryApi } from "@reduxjs/toolkit/query/react";
import { type CreateTRPCClientOptions, type CreateTRPCProxyClient } from "@trpc/client";
import { type AnyRouter } from "@trpc/server";

/**
 * Client specific options when creating trpc api. You can either
 *  - pass in already created proxy client
 *  - pass in client options, so client is created for you
 *  - pass in callback that gets access to BaseQueryApi from RTK. This is useful for example when you need to access redux store when creating the client.
 **/
export type CreateTRPCApiClientOptions<TRouter extends AnyRouter> =
  | {
      client: CreateTRPCProxyClient<TRouter>;
      clientOptions?: never;
      getClient?: never;
    }
  | {
      client?: never;
      clientOptions: CreateTRPCClientOptions<TRouter>;
      getClient?: never;
    }
  | {
      client?: never;
      clientOptions?: never;
      getClient: (
        baseQueryApi: BaseQueryApi,
      ) => Promise<CreateTRPCProxyClient<TRouter>>;
    };
