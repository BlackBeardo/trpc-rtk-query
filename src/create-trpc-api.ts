import type {
  BaseQueryFn,
  MutationDefinition,
  QueryDefinition,
} from "@reduxjs/toolkit/query/react";
import type { CreateTRPCClientOptions } from "@trpc/client";
import type {
  AnyProcedure,
  AnyRouter,
  Procedure,
  inferProcedureInput,
  inferProcedureOutput,
} from "@trpc/server";

import { createApi } from "@reduxjs/toolkit/query/react";
import { createTRPCUntypedClient } from "@trpc/client";

// Follows trpc internal infer type pattern
type inferProcedureType<TProcedure extends AnyProcedure> = TProcedure extends Procedure<
  infer ProcedureType,
  // any is okay here, we don't care about the second param
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  any
>
  ? ProcedureType
  : never;

// TODO: handle deep merged routes with namespaces, like user.procedure
type CreateTRPCApiEndpointDefinitions<
  TRouter extends AnyRouter,
  BaseQuery extends BaseQueryFn,
  TagTypes extends string,
  ReducerPath extends string,
  Procedures = TRouter["_def"]["record"],
> = {
  [EndpointName in keyof Procedures]: Procedures[EndpointName] extends AnyProcedure
    ? inferProcedureType<Procedures[EndpointName]> extends infer ProcedureType
      ? ProcedureType extends "query"
        ? QueryDefinition<
            inferProcedureInput<Procedures[EndpointName]>,
            BaseQuery,
            TagTypes,
            inferProcedureOutput<Procedures[EndpointName]>,
            ReducerPath
          >
        : ProcedureType extends "mutation"
        ? MutationDefinition<
            inferProcedureInput<Procedures[EndpointName]>,
            BaseQuery,
            TagTypes,
            inferProcedureOutput<Procedures[EndpointName]>,
            ReducerPath
          >
        : never
      : never
    : never;
};

const deCapitalize = (string_: string) => {
  const firstChar = string_[0];
  return firstChar ? string_.replace(firstChar, firstChar?.toLowerCase()) : string_;
};

export const createTRPCApi = <TRouter extends AnyRouter>(
  options: CreateTRPCClientOptions<TRouter>,
) => {
  // TRPC Client
  // TODO: Allow passing from outside
  const client = createTRPCUntypedClient(options);

  // RTK Query api
  // TODO: allow passing original api from outside and inject endpoints instead of
  // generating the api from scratch
  const baseQuery = async ({
    procedureArgs,
    procedureName,
    procedureType,
  }: {
    procedureArgs: unknown;
    procedureName: string;
    procedureType: "mutation" | "query";
  }) => {
    try {
      return {
        data: await client[procedureType](procedureName, procedureArgs),
      };
    } catch (error) {
      // TODO: proper error typings for query results (i.e. TRPCClientError)
      // TODO: Try to catch non-serializable errors here, and turn them serializable
      console.log("error", error);
      return { error: "Error when calling client" };
    }
  };
  const reducerPath = "TRPCApi" as const;
  type BaseQuery = typeof baseQuery;
  type TagTypes = never; // No tags
  type ReducerPath = typeof reducerPath;
  // Create underlying api that can be proxyed
  const nonProxyApi = createApi<
    BaseQuery,
    CreateTRPCApiEndpointDefinitions<TRouter, BaseQuery, TagTypes, ReducerPath>,
    ReducerPath,
    TagTypes
  >({
    baseQuery,
    // We're injecting endpoints later when they're exported
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    endpoints: () => ({} as any),
    reducerPath,
  });

  // TODO:
  // add support for
  // - deep properties such as api.endpoints.getPosts.useQuerySubscription
  const regexesWithProcedureType = [
    {
      procedureType: "query",
      regex: /use(\w+)Query/,
    },
    {
      procedureType: "query",
      regex: /useLazy(\w+)Query/,
    },
    {
      procedureType: "mutation",
      regex: /use(\w+)Mutation/,
    },
  ] as const;

  return new Proxy(nonProxyApi, {
    get(target, property) {
      // === Always call the property if it's already defined

      // eslint-disable-next-line no-prototype-builtins
      if (target.hasOwnProperty(property)) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return (target as any)[property as any];
      }

      // === Othewise, try to generate property

      // Can't really do anything here if symbol is not already part of api
      if (typeof property === "symbol") {
        throw new TypeError("Calling api with new symbol properties is not supported");
      }

      for (const { procedureType, regex } of regexesWithProcedureType) {
        const match = regex.exec(property);
        if (!match) {
          continue;
        }
        // matched group is the 2nd arg
        const capitalizedEndpointName = match[1];
        // pass through if parsing not okay with this regex
        if (!capitalizedEndpointName) {
          continue;
        }
        const procedure = deCapitalize(capitalizedEndpointName);
        target.injectEndpoints({
          endpoints: (builder) => ({
            [procedure]: builder[procedureType]({
              query: (procedureArguments: unknown) => ({
                procedureArgs: procedureArguments,
                procedureName: procedure,
                procedureType: "query",
              }),
            }),
          }),
        });

        // Return newly generated property
        return target[property as keyof typeof target];
      }

      throw new TypeError(
        `Property ${property} is not defined and could not be generated`,
      );
    },
  });
};
