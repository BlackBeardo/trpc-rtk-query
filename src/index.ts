export { createTRPCApi } from "./create-trpc-api";
