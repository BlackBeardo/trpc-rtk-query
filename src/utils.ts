import type { AnyProcedure, Procedure } from "@trpc/server";

// Follows trpc internal infer type patterns
export type inferProcedureType<TProcedure extends AnyProcedure> =
  TProcedure extends Procedure<
    infer ProcedureType,
    // any is okay here, we don't care about the second param
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    any
  >
    ? ProcedureType
    : never;

// Primitive type from type-fest
type Primitive = bigint | boolean | null | number | string | symbol | undefined;

// TODO: Decide what kind of paths we want to have for deeply nested routes
type PathToString<Path extends string[]> = Path extends [infer first extends string]
  ? first
  : Path extends [infer first extends string, ...infer rest extends string[]]
  ? `${first}_${PathToString<rest>}`
  : ``;

type FlattenPairs<Value, Path extends string[] = []> = {
  [Key in keyof Value]: Value[Key] extends Primitive
    ? [PathToString<[...Path, Extract<Key, string>]>, Value[Key]]
    : FlattenPairs<Value[Key], [...Path, Extract<Key, string>]>;
}[keyof Value];

// TODO: fix possibly infinite type instantiation
type Flatten<Value> = { [Pair in FlattenPairs<Value> as Pair[0]]: Pair[1] };

type A = {
  getUserById: number;
  listUsers: string;
  nested: {
    deep: {
      getVeryNestedMessage: boolean;
    };
  };
};

type B = Flatten<A>;
