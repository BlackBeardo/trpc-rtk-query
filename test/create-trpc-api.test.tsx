import { configureStore } from "@reduxjs/toolkit";
import { type BaseQueryApi, createApi } from "@reduxjs/toolkit/query/react";
import { createTRPCProxyClient } from "@trpc/client";
import { setTimeout } from "node:timers/promises";
import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { beforeAll, describe, expect, it, vi } from "vitest";

import { createTRPCApi, injectTRPCEndpointsToApi } from "../src/create-trpc-api";
import {
  type AppRouter,
  startTestServer,
  testClientOptions,
  userFixtures,
} from "./fixtures";

describe("create-trpc-api", () => {
  // Client side options that can be passed in when testing
  const existingApiTestQueryArgument = "giveMeData";
  const existingApiTestQuerySuccessResponse = {
    returning: "data",
  };
  const existingApiTestQueryFailureResponse = {
    error: "bad stuff happened",
  };
  // helper so we can always generate new api instance when needed
  const createApiLazily = () =>
    createApi({
      baseQuery: (baseQueryArguments: { getResponseArgument: string }) => {
        if (baseQueryArguments.getResponseArgument !== existingApiTestQueryArgument) {
          return existingApiTestQueryFailureResponse;
        }
        return {
          data: existingApiTestQuerySuccessResponse,
        };
      },
      endpoints: (builder) => ({
        getResponse: builder.query<string, string>({
          query: (getResponseArgument: string) => ({
            getResponseArgument,
          }),
        }),
      }),
    });

  describe.each([
    {
      testCase: "using api generated by createTRPCApi",
    },
    {
      apiForCreateApiOptions: {
        createApiLazily,
      },
      testCase: "using pre made api",
    },
  ])("when $testCase and ", ({ apiForCreateApiOptions }) => {
    describe.each([
      {
        getApi: () => {
          const base = { clientOptions: testClientOptions };
          if (apiForCreateApiOptions?.createApiLazily) {
            const existingApi = createApiLazily();
            return injectTRPCEndpointsToApi<AppRouter, typeof existingApi>({
              ...base,
              existingApi,
            });
          }
          return createTRPCApi<AppRouter>(base);
        },
        testCase: "creating client from clientOptions",
      },
      {
        getApi: () => {
          const base = {
            client: createTRPCProxyClient<AppRouter>(testClientOptions),
          };
          if (apiForCreateApiOptions?.createApiLazily) {
            const existingApi = createApiLazily();
            return injectTRPCEndpointsToApi<AppRouter, typeof existingApi>({
              ...base,
              existingApi,
            });
          }
          return createTRPCApi(base);
        },
        testCase: "using passed client",
      },
      {
        getApi: () => {
          const base = {
            getClient: async (baseQueryApi: BaseQueryApi) => {
              // Check that correct apiArgs object is passed
              expect(baseQueryApi.type).toBeDefined();
              expect(baseQueryApi.endpoint).toBeDefined();
              // Return proxy client
              return createTRPCProxyClient<AppRouter>(testClientOptions);
            },
          };
          if (apiForCreateApiOptions?.createApiLazily) {
            const existingApi = createApiLazily();
            return injectTRPCEndpointsToApi<AppRouter, typeof existingApi>({
              ...base,
              existingApi,
            });
          }
          return createTRPCApi(base);
        },
        testCase: "using getClient to get the client",
      },
    ])("when $testCase", ({ getApi }) => {
      it("Generates an api instance", () => {
        const api = getApi();
        expect(api).toBeDefined();
      });
      it("Generates queries ", () => {
        const api = getApi();
        const { useGetUserByIdQuery, useListUsersQuery } = api;
        expect(useGetUserByIdQuery).toBeDefined();
        expect(useListUsersQuery).toBeDefined();
      });
      it("Generates mutations with correct typings", () => {
        const api = getApi();
        const { useCreateUserMutation, useUpdateNameMutation } = api;
        expect(useUpdateNameMutation).toBeDefined();
        expect(useCreateUserMutation).toBeDefined();
      });
      it.each([
        "useQuery",
        "useQueryState",
        "useQuerySubscription",
        "useLazyQuery",
        "useLazyQuerySubscription",
      ] as const)(
        "Generates %s hook when accessing hooks through endpoints[endpoint] property",
        (queryName) => {
          const api = getApi();
          const query = api.endpoints.getUserById[queryName];
          expect(query).toBeDefined();
        },
      );
      it("Generates usePrefetch", () => {
        const api = getApi();
        const { usePrefetch } = api;
        expect(usePrefetch).toBeDefined();
      });

      it("Generates hook for deeply nested route", () => {
        const api = getApi();
        const { useNested_Deep_GetVeryNestedMessageQuery } = api;
        expect(useNested_Deep_GetVeryNestedMessageQuery).toBeDefined();
      });

      it("Generates hook for deeply nested routes through endpoints[endpoint]", () => {
        const api = getApi();
        const query =
          api.endpoints.nested_Deep_GetVeryNestedMessage.useQuerySubscription;
        expect(query).toBeDefined();
      });
      describe("and making actual requests with hooks renders correctly", () => {
        // render component to json for snapshots
        // eslint-disable-next-line unicorn/consistent-function-scoping
        const renderedToJSon = (component: renderer.ReactTestRenderer) => {
          const result = component.toJSON();
          expect(result).toBeDefined();
          expect(result).not.toBeInstanceOf(Array);
          return result as renderer.ReactTestRendererJSON;
        };

        // Keep this in scope for easier type deriving
        // eslint-disable-next-line unicorn/consistent-function-scoping
        const createReactTestApp = (getApiForStore: typeof getApi) => {
          const api = getApiForStore();
          const store = configureStore({
            middleware: (getDefaultMiddleware) => [
              ...getDefaultMiddleware(),
              api.middleware,
            ],
            reducer: {
              [api.reducerPath]: api.reducer,
            },
          });

          const createComponentWrapper = (Component: () => React.JSX.Element) =>
            renderer.create(
              <Provider store={store}>
                <Component />
              </Provider>,
            );
          return {
            api,
            createComponentWrapper,
            store,
          };
        };

        // how much to wait for loading state to resolve
        // TODO: fix with using actual rerender with waitFor from testing-library
        const msToWaitBeforeRenderingWithoutLoadingState = 150;
        beforeAll(async () => {
          const { close } = await startTestServer();
          console.log("waiting a bit before running integration tests");
          await setTimeout(250);
          return () => close();
        });
        it("with successful useUserIdQuery", async () => {
          const { api, createComponentWrapper } = createReactTestApp(getApi);
          const Component = () => {
            const { useGetUserByIdQuery } = api;
            const userId = 1;
            const { data, error, isLoading } = useGetUserByIdQuery(userId);
            if (isLoading) {
              return <div>Loading...</div>;
            }
            if (error || !data) {
              return <div>Error</div>;
            }
            expect(data).toEqual(userFixtures[1]);
            return (
              <div>
                <p>Id: {data.id}</p>
                <p>Name: {data.name}</p>
              </div>
            );
          };
          const app = createComponentWrapper(Component);
          // first render
          let result = renderedToJSon(app);
          await setTimeout(msToWaitBeforeRenderingWithoutLoadingState);
          expect(result).toMatchSnapshot();
          // This render needs a bit more time for getClient case, I guess because of
          // forming queryClient takes a bit more time
          // TODO: plz fix with waitFor or something
          // result after data has loaded and component has re-rendered
          result = renderedToJSon(app);
          expect(JSON.stringify(result)).not.toContain("Loading...");
          expect(JSON.stringify(result)).not.toContain("Error");
          expect(result).toMatchSnapshot();
        });
        it("with failing useUserIdQuery", async () => {
          const { api, createComponentWrapper } = createReactTestApp(getApi);
          const Component = () => {
            const { useGetUserByIdQuery } = api;
            const userId = 4;
            // TODO: errors should be properly typed from basequery!
            const { data, error, isLoading } = useGetUserByIdQuery(userId);
            if (isLoading) {
              return <div>Loading...</div>;
            }
            if (error || !data) {
              return <div>Error</div>;
            }
            expect(data).toEqual(userFixtures[1]);
            return (
              <div>
                <p>Id: {data.id}</p>
                <p>Name: {data.name}</p>
              </div>
            );
          };
          const app = createComponentWrapper(Component);
          // first render
          let result = renderedToJSon(app);
          expect(result).toMatchSnapshot();
          await setTimeout(msToWaitBeforeRenderingWithoutLoadingState);
          // result after data has loaded and component has re-rendered
          result = renderedToJSon(app);
          expect(JSON.stringify(result)).not.toContain("Loading...");
          expect(JSON.stringify(result)).toContain("Error");
          expect(result).toMatchSnapshot();
        });
        it("with successful deep nested query", async () => {
          const { api, createComponentWrapper } = createReactTestApp(getApi);
          const Component = () => {
            const { useNested_Deep_GetVeryNestedMessageQuery } = api;
            const myInput = "heyoooo";
            const { data, error, isLoading } = useNested_Deep_GetVeryNestedMessageQuery(
              {
                deepInput: myInput,
              },
            );
            if (isLoading) {
              return <div>Loading...</div>;
            }
            if (error || !data) {
              return <div>Error</div>;
            }
            expect(data.inputBack).toStrictEqual(myInput);
            return (
              <div>
                <p>inputBack: {data.inputBack}</p>
                <p>messageFromDeep: {data.messageFromDeep}</p>
              </div>
            );
          };
          const app = createComponentWrapper(Component);
          // first render
          let result = renderedToJSon(app);
          expect(result).toMatchSnapshot();
          await setTimeout(msToWaitBeforeRenderingWithoutLoadingState);
          // result after data has loaded and component has re-rendered
          result = renderedToJSon(app);
          expect(JSON.stringify(result)).not.toContain("Loading...");
          expect(JSON.stringify(result)).not.toContain("Error");
          expect(result).toMatchSnapshot();
        });
        it("with call to usePrefetch", async () => {
          const { api, createComponentWrapper } = createReactTestApp(getApi);
          const userId = 1;
          const PrefetcherComponent = () => {
            const { usePrefetch } = api;
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore ignore for now, for runtime testing. Fix type later
            const prefetch = usePrefetch("getUserById");
            expect(prefetch).toBeDefined();
            prefetch(userId);
            return <>prefetched</>;
          };
          const QueryComponent = () => {
            const { useQueryState } = api.endpoints.getUserById;
            const { data } = useQueryState(userId);
            expect(data).toEqual(userFixtures[1]);
            return (
              <div>
                <p>Id: {data!.id}</p>
                <p>Name: {data!.name}</p>
              </div>
            );
          };
          const Parent = () => {
            const [showQueryComponent, setShowQueryComponent] = React.useState(false);
            const handleMouseEnter = () => {
              setShowQueryComponent(true);
            };
            return (
              <div onMouseEnter={handleMouseEnter}>
                {showQueryComponent ? <QueryComponent /> : <PrefetcherComponent />}
              </div>
            );
          };
          // render prefetcher component
          const component = createComponentWrapper(Parent);
          let result = renderedToJSon(component);
          expect(JSON.stringify(result)).toContain("prefetched");
          expect(result).toMatchSnapshot();
          // wait for data to load
          await setTimeout(msToWaitBeforeRenderingWithoutLoadingState);
          // manually trigger the callback
          result.props["onMouseEnter"]();
          // render query component and check that prefetch was actually called
          // properly through proxy (e.g. query state should be loaded)
          result = renderedToJSon(component);
          expect(JSON.stringify(result)).not.toContain("Error");
          expect(JSON.stringify(result)).toContain("Id");
          expect(JSON.stringify(result)).toContain("Name");
          expect(result).toMatchSnapshot();
        });
        it("does not inject endpoints again", async () => {
          const { api, createComponentWrapper } = createReactTestApp(getApi);
          // Try to listen error log from rtk
          const consoleErrorMock = vi
            .spyOn(console, "error")
            // eslint-disable-next-line @typescript-eslint/no-empty-function
            .mockImplementation(() => {});
          const FirstComponent = () => {
            const { useListUsersQuery } = api;
            const { data } = useListUsersQuery();
            return <>{JSON.stringify(data)}</>;
          };
          const SecondComponent = () => {
            const { useListUsersQuery } = api;
            const { data } = useListUsersQuery();
            return <>{JSON.stringify(data)}</>;
          };
          const Parent = () => {
            return (
              <>
                <FirstComponent />
                <SecondComponent />
              </>
            );
          };
          const app = createComponentWrapper(Parent);
          // first render
          renderedToJSon(app);
          expect(consoleErrorMock).not.toHaveBeenCalled();
        });
      });
      it("Can be extended with enchangeEndpoints", () => {
        const api = getApi();
        expect(api.enhanceEndpoints).toBeDefined();
        // This is here just to see that it can be executed
        api.enhanceEndpoints({
          addTagTypes: ["User"],
          endpoints: {
            getUserById: {
              // TODO: actually test this works
              // eslint-disable-next-line @typescript-eslint/no-empty-function
              onQueryStarted() {},
              providesTags: ["User"],
            },
            updateName: {
              // TODO: actually test this works
              // eslint-disable-next-line @typescript-eslint/no-empty-function
              onQueryStarted() {},
            },
          },
        });
      });
    });
  });
  it("doesn't replace previous hooks when passing in and existing api", () => {
    const base = { clientOptions: testClientOptions };
    const existingApi = createApiLazily();
    const api = injectTRPCEndpointsToApi<AppRouter, typeof existingApi>({
      ...base,
      existingApi,
    });
    const { useListUsersQuery } = api;
    expect(useListUsersQuery).toBeDefined();
    const { useGetResponseQuery } = api;
    expect(useGetResponseQuery).toBeDefined();
  });
});
