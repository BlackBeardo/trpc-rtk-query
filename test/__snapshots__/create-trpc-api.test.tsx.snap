// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`making actual requests > with successful useUserIdQuery 1`] = `
<div>
  Loading...
</div>
`;

exports[`making actual requests > with successful useUserIdQuery 2`] = `
<div>
  <p>
    Id: 
    1
  </p>
  <p>
    Name: 
    John Johnny
  </p>
</div>
`;

exports[`making actual requests with hooks renders correctly > with failing useUserIdQuery 1`] = `
<div>
  Loading...
</div>
`;

exports[`making actual requests with hooks renders correctly > with failing useUserIdQuery 2`] = `
<div>
  Error
</div>
`;

exports[`making actual requests with hooks renders correctly > with successful useUserIdQuery 1`] = `
<div>
  Loading...
</div>
`;

exports[`making actual requests with hooks renders correctly > with successful useUserIdQuery 2`] = `
<div>
  <p>
    Id: 
    1
  </p>
  <p>
    Name: 
    John Johnny
  </p>
</div>
`;
